R-package/tests/testthat/test_basic.R

```diff
@@ -188,6 +188,60 @@ test_that("lightgbm() rejects negative or 0 value passed to nrounds", {
   }
 })
 
+test_that("lightgbm() accepts nrounds as either a top-level argument or parameter", {
+  nrounds <- 15L
+
+  set.seed(708L)
+  top_level_bst <- lightgbm(
+    data = train$data
+    , label = train$label
+    , nrounds = nrounds
+    , params = list(
+      objective = "regression"
+      , metric = "l2"
+      , num_leaves = 5L
+    )
+    , save_name = tempfile(fileext = ".model")
+  )
+
+  set.seed(708L)
+  param_bst <- lightgbm(
+    data = train$data
+    , label = train$label
+    , params = list(
+      objective = "regression"
+      , metric = "l2"
+      , num_leaves = 5L
+      , nrounds = nrounds
+    )
+    , save_name = tempfile(fileext = ".model")
+  )
+
+  set.seed(708L)
+  both_customized <- lightgbm(
+    data = train$data
+    , label = train$label
+    , nrounds = 20L
+    , params = list(
+      objective = "regression"
+      , metric = "l2"
+      , num_leaves = 5L
+      , nrounds = nrounds
+    )
+    , save_name = tempfile(fileext = ".model")
+  )
+
+  expect_equal(param_bst$current_iter(), top_level_bst$current_iter())
+  expect_equal(param_bst$best_score
+               , top_level_bst$best_score
+               , tolerance = TOLERANCE)
+
+  expect_equal(param_bst$current_iter(), both_customized$current_iter())
+  expect_equal(param_bst$best_score
+               , both_customized$best_score
+               , tolerance = TOLERANCE)
```

Comment:
I believe we should check here for the exact equality without any `tolerance`. Given the fixed environment and the same set of parameters, consecutive calls of lightgbm should produce bit-by-bit identical results. Just like here:
https://github.com/microsoft/LightGBM/blob/d62378b45314c2688deef7333ae16fcc9863d881/tests/python_package_test/test_basic.py#L66-L67

Reply:
I agree with @StrikerRUS . And just to be clear, that would mean using `testthat::expect_identical()`, not just removing `tolerance`.

Since `expect_equal()` already allows for small differences by default: https://rdrr.io/cran/testthat/man/equality-expectations.html

Reply:
I swapped this to `expect_identical`. When running interactively, this test was non-deterministic; it would sometimes fail with message ``top_level_l2` not identical to `params_l2`. Objects equal but not identical.` and sometimes succeed. My attempts to capture this using reprex succeeded every time, so I'm not sure if something about the interactive environment, R Studio, or (most likely) the way I was running tests was causing the issue, but I figured I'd flag all the same.